components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/dao/impl/IdentityUserMetadataMgtDAOImpl.java

```diff
@@ -0,0 +1,508 @@
+                                " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
+                } else if (!IdentityUtil.isUseCaseSensitiveUsernameForCacheKeys(userStoreManager)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Case insensitive username for cache key is used. Changing username from : "
+                                + userName + " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
+                }
+
+                String domainName = userStoreManager.getRealmConfiguration().
+                        getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME);
+
+                IdentityDataStoreCacheKey key = new IdentityDataStoreCacheKey(domainName, userName);
+                int tenantId = userStoreManager.getTenantId();
+                UserIdentityClaim userIdentityDTO = identityDataStoreCache.getValueFromCache(key, tenantId);
+
+                if (userIdentityDTO != null && log.isDebugEnabled()) {
+                    StringBuilder data = new StringBuilder("{");
+                    if (userIdentityDTO.getUserIdentityDataMap() != null) {
+                        for (Map.Entry<String, String> entry : userIdentityDTO.getUserIdentityDataMap().entrySet()) {
+                            data.append("[").append(entry.getKey()).append(" = ").append(entry.getValue())
+                                    .append("], ");
+                        }
+                    }
+                    if (data.indexOf(",") >= 0) {
+                        data.deleteCharAt(data.lastIndexOf(","));
+                    }
+                    data.append("}");
+                    log.debug("Loaded UserIdentityClaimsDO from cache for user :" + userName + " with claims: " + data);
+
+                }
+                return userIdentityDTO;
+            }
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            log.error("Error while obtaining tenant ID from user store manager");
+        }
+        return null;
+    }
+
+    @Override
+    public void storeUserMetadataToCache(UserStoreManager userStoreManager, UserIdentityClaim userIdentityDTO) {
+
+        try {
+            if (userIdentityDTO != null && userIdentityDTO.getUserName() != null) {
+                String userName = UserCoreUtil.removeDomainFromName(userIdentityDTO.getUserName());
+                if (!IdentityUtil.isUserStoreCaseSensitive(userStoreManager)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Case insensitive user store found. Changing username from : " + userName +
+                                " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
```

Comment:
This also can be moved out of the if block to avoid calling repeatedly. Please check for other applicable cases as well.

Reply:
Fixed in https://github.com/wso2-extensions/identity-governance/pull/697/commits/dd3419fd69cc40e847e40ed9dfb29c867a3f0018